numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -4229,6 +4231,144 @@ _reload_guard(PyObject *NPY_UNUSED(self)) {
 }
 
 
+NPY_NO_EXPORT PyObject *
+from_dlpack(PyObject *NPY_UNUSED(self), PyObject *obj) {
+    PyObject *capsule = PyObject_CallMethod((PyObject *)obj->ob_type,
+            "__dlpack__", "O", obj);
+    if (capsule == NULL) {
+        return NULL;
+    }
+
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(capsule,
+        NPY_DLPACK_CAPSULE_NAME);
+
+    if (managed == NULL) {
+        Py_XDECREF(capsule);
+        return NULL;
+    }
+
+    const int ndim = managed->dl_tensor.ndim;
+    if (ndim >= NPY_MAXDIMS) {
+        PyErr_SetString(PyExc_RuntimeError,
+                "maxdims of DLPack tensor is higher than the supported "
+                "maxdims.");
+        Py_XDECREF(capsule);
+        return NULL;
+    }
+
+    if (managed->dl_tensor.device.device_type != kDLCPU &&
+            managed->dl_tensor.device.device_type != kDLCUDAHost) {
```

Comment:
A supported device type list is this: `kDLCPU`, `kDLCUDAHost`, `kDLROCMHost`, `kDLCUDAManaged`. They are all readable/writable on CPU.

Reply:
Ah, sorry, the last two are new (https://github.com/dmlc/dlpack/pull/71) and not yet included in the bundled header. Let's add them in a separate PR.

Reply:
I have updated the bundled header, and applied this suggestion.